x-pack/plugins/security_solution/server/usage/detections/rules/get_metrics.ts

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { ElasticsearchClient, SavedObjectsClientContract, Logger } from 'kibana/server';
+import type { RuleAdoption } from './types';
+
+import { updateRuleUsage } from './update_usage';
+import { getDetectionRules } from '../../queries/get_detection_rules';
+import { getAlerts } from '../../queries/get_alerts';
+import { MAX_PER_PAGE, MAX_RESULTS_WINDOW } from '../../constants';
+import { getInitialRulesUsage } from './get_initial_usage';
+import { getCaseComments } from '../../queries/get_case_comments';
+import { getRuleIdToCasesMap } from './transform_utils/get_rule_id_to_cases_map';
+import { getAlertIdToCountMap } from './transform_utils/get_alert_id_to_count_map';
+import { getRuleIdToEnabledMap } from './transform_utils/get_rule_id_to_enabled_map';
+import { getRuleObjectCorrelations } from './transform_utils/get_rule_object_correlations';
+
+// eslint-disable-next-line no-restricted-imports
+import { legacyGetRuleActions } from '../../queries/legacy_get_rule_actions';
+
+export interface GetRuleMetricsOptions {
+  kibanaIndex: string;
+  signalsIndex: string;
+  esClient: ElasticsearchClient;
+  savedObjectsClient: SavedObjectsClientContract;
+  logger: Logger;
+}
+
+export const getRuleMetrics = async ({
+  kibanaIndex,
+  signalsIndex,
+  esClient,
+  savedObjectsClient,
+  logger,
+}: GetRuleMetricsOptions): Promise<RuleAdoption> => {
+  try {
+    // gets rule saved objects
+    const ruleResults = await getDetectionRules({
+      esClient,
+      kibanaIndex,
+      maxPerPage: MAX_PER_PAGE,
+      maxSize: MAX_RESULTS_WINDOW,
+      logger,
+    });
+
+    // early return if we don't have any detection rules then there is no need to query anything else
+    if (ruleResults.length === 0) {
+      return {
+        detection_rule_detail: [],
+        detection_rule_usage: getInitialRulesUsage(),
+      };
+    }
+
+    // gets the alerts data objects
+    const detectionAlertsResp = await getAlerts({
+      esClient,
+      signalsIndex: `${signalsIndex}*`,
+      maxPerPage: MAX_PER_PAGE,
+      maxSize: MAX_RESULTS_WINDOW,
+      logger,
+    });
+
+    // gets cases saved objects
+    const caseComments = await getCaseComments({
+      savedObjectsClient,
+      maxSize: MAX_PER_PAGE,
+      maxPerPage: MAX_RESULTS_WINDOW,
+      logger,
+    });
+
+    // gets the legacy rule actions to track legacy notifications.
+    const legacyRuleActions = await legacyGetRuleActions({
+      savedObjectsClient,
+      maxSize: MAX_PER_PAGE,
+      maxPerPage: MAX_RESULTS_WINDOW,
+      logger,
+    });
```

Comment:
It looks like we could `Promise.all` these calls.

Reply:
Yep, good 👁️ , I added a Promise.all() to all of them but the first call so we could short circuit on the first one if possible before the others.